cpp/src/table/row_operators.cu

```diff
@@ -39,33 +87,57 @@ namespace {
  * non-decomposed table, which are pruned during decomposition.
  *
  * For example, if the original table has a column `Struct<Struct<int, float>, decimal>`,
+ *
  *      S1
  *     / \
  *    S2  d
  *   / \
  *  i   f
+ *
  * then after decomposition, we get three columns:
  * `Struct<Struct<int>>`, `float`, and `decimal`.
- * 0   2   1  <- depths
- * S1
- * |
- * S2      d
- * |
- * i   f
+ *
+ *  0   2   1  <- depths
+ *  S1
+ *  |
+ *  S2      d
+ *  |
+ *  i   f
+ *
  * The depth of the first column is 0 because it contains all its parent levels, while the depth
  * of the second column is 2 because two of its parent struct levels were pruned.
  *
- * Similarly, a struct column of type Struct<int<Struct<float, decimal>> is decomposed as follows
+ * Similarly, a struct column of type Struct<int, Struct<float, decimal>> is decomposed as follows
+ *
  *     S1
  *    / \
  *   i   S2
  *      / \
  *     f   d
  *
- * 0   1   2  <- depths
- * S1  S2  d
- * |   |
- * i   f
+ *  0   1   2  <- depths
+ *  S1  S2  d
+ *  |   |
+ *  i   f
+ *
+ * When list columns are present, The decomposition is performed similarly to pure structs but list
```

Comment:
I want to make sure that I understand this. Pruning of struct parents is safe because once you've compared it once, you are guaranteed that the struct "looks" the same at the root of each of the verticalized trees (it has the same children) so you don't need to check it again, is that right? For lists, this doesn't work because unlike with structs you need to compare all of the elements for equality, so all you can do is verticalize the struct children while leaving the lists at the root of each verticalized tree?

It would be helpful to add an explanation here as well for posterity!

Reply:
That is correct, also with lists, you want the parent levels because there is no other zero copy way to expand your range selection until the leaf. Pruning list levels will make the size of one decomposed branch different than the other.